README.md

```diff
@@ -1,130 +1,214 @@
+
+Second, if you use Smartcar hosting you must provide an `onComplete` function with the following signature `onComplete(error, code, [state])` to the Smartcar constructor. This `onComplete` method will be called when the user completes the authorization flow. If they approved access you'll need to exchange the `code` parameter for an access token. If they denied access you'll need to handle the error (passed in the `error` argument).
+
+If you host the redirect yourself you can provide any valid redirect URI (see the [Register](https://smartcar.com/docs#redirect-uris) section of our docs) and `onComplete` is optional. You may source `redirect.js` in your redirect popup to close out the window and trigger firing of the `onComplete` (if you've provided one to the `Smartcar` constructor).
+
+Below are example flows for each method.
+
+### Example Flows
+
+#### Smartcar-hosted
 
 1. User clicks "Connect your car" button (or similar) on your application's website.
-2. The user is redirected to a new page, either as a popup (if using the `openDialog` and/or `addClickHandler` methods), or in the same page (if the client redirects to the link generated by `generateLink` directly, without using `openDialog` and/or `addClickHandler`). This page requires the user authenticate with their vehicle credentials.
-3. The user will be asked to authorize your application to connect to their vehicle.
-4. Smartcar's services will redirect back to your application's redirectUri. The redirect will include an authorization code if the user approved the authorization, an error otherwise. An optional callback function (passed in as `onComplete`) will be invoked if `callback.js` is imported in the page served at your application's redirectUri.
+2. User is redirected to an OAuth authentication page, using the `openDialog` or `addClickHandler` methods. This page requires the user to authenticate with their vehicle credentials.
+
+    *Note:* due to the implementation of `redirect.js`, the redirect page must be opened in a separate window.
+3. After entering credentials the user will be presented with a list of their vehicles to authorize and a selection of permissions to grant your app.
+4. Once the user has selected their vehicle & selected permissions, this SDK will automagically trigger the `onComplete` function you provide with the authorization code or an error. If called with a code you must send this code to your backend to then exchange with Smartcar for an access token.
```

Comment:
It won't display an error, it'll call the function with an error (an object that contains the description of the error so that the developer can handle it in some way, like redirecting the user to a new web page). Even with that I think it's still pretty awkward if you have any improvements!

Reply:
How about:
4. Once the user has selected their vehicle and selected permissions, this SDK will either automagically trigger the 'onComplete' function that you provide with the authorization code, or it will **produce** an error. If **the 'onComplete' function is called**, you must send **the authorization** code to your back end to then exchange it with Smartcar for an access token.